File: src/apps/Andriod/app/src/main/java/com/dkellycollins/triggerfinger/managers/entity/impl/GameStateManager.java
package com.dkellycollins.triggerfinger.managers.entity.impl;

import com.dkellycollins.triggerfinger.data.daos.IGameStateDao;
import com.dkellycollins.triggerfinger.data.entity.IGameState;
import com.dkellycollins.triggerfinger.managers.entity.IGameStateManager;
import com.dkellycollins.triggerfinger.util.qa.Assert;

public class GameStateManager implements IGameStateManager {

    private IGameStateDao _dao;

    public GameStateManager(IGameStateDao dao) {
        Assert.isNotNull(dao, "dao");

        _dao = dao;
    }

    @Override
    public IGameState retrieve() {
        return _dao.retrieve();
    }

    @Override
    public void TogglePause() {
        IGameState gameState = retrieve();

        _dao.update(gameState.IsGameOver(), !gameState.IsPaused());
    }

    @Override
    public void TogglePause(boolean isPaused) {
        IGameState gameState = retrieve();

        _dao.update(gameState.IsGameOver(), isPaused);
    }

    @Override
    public void SetGameOver() {
        IGameState gameState = retrieve();

        _dao.update(true, gameState.IsPaused());
    }

    @Override
    public void Reset() {
        _dao.update(false, false);
    }
}

File: src/apps/Android-Kotlin/app/src/main/kotlin/com/dkellycollins/triggerfinger/managers/entity/impl/GameStateManager.kt
package com.dkellycollins.triggerfinger.managers.entity.impl

import com.dkellycollins.triggerfinger.data.daos.IGameStateDao
import com.dkellycollins.triggerfinger.data.entity.IGameState
import com.dkellycollins.triggerfinger.managers.entity.IGameStateManager

class GameStateManager(private val _dao: IGameStateDao) : IGameStateManager {

    override fun retrieve(): IGameState {
        return _dao.retrieve()
    }

    override fun TogglePause() {
        val gameState = retrieve()

        _dao.update(gameState.IsGameOver(), !gameState.IsPaused())
    }

    override fun TogglePause(isPaused: Boolean) {
        val gameState = retrieve()

        _dao.update(gameState.IsGameOver(), isPaused)
    }

    override fun SetGameOver() {
        val gameState = retrieve()

        _dao.update(true, gameState.IsPaused())
    }

    override fun Reset() {
        _dao.update(false, false)
    }
}
